server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java

```diff
@@ -81,74 +84,100 @@ public void define(WebService.NewController context) {
+      .setSince("6.4");
   }
 
   @Override
   public void handle(Request wsRequest, Response wsResponse) throws Exception {
-    SuggestionsWsResponse searchWsResponse = doHandle(wsRequest.param(URL_PARAM_QUERY));
-    writeProtobuf(searchWsResponse, wsRequest, wsResponse);
-  }
+    String query = wsRequest.param(URL_PARAM_QUERY);
+    String more = wsRequest.param(URL_PARAM_MORE);
 
-  SuggestionsWsResponse doHandle(String query) {
-    List<Qualifier> resultsPerQualifier = getResultsOfAllQualifiers(query);
+    ComponentIndexQuery.Builder queryBuilder = ComponentIndexQuery.builder().setQuery(query);
 
-    return SuggestionsWsResponse.newBuilder()
-      .addAllResults(resultsPerQualifier)
-      .build();
-  }
+    List<ComponentsPerQualifier> componentsPerQualifiers = getComponentsPerQualifiers(more, queryBuilder);
+    String warning = getWarning(query);
 
-  private List<Qualifier> getResultsOfAllQualifiers(String query) {
-    ComponentIndexQuery componentIndexQuery = new ComponentIndexQuery(query)
-      .setQualifiers(Arrays.asList(QUALIFIERS))
-      .setLimit(NUMBER_OF_RESULTS_PER_QUALIFIER);
-
-    List<ComponentsPerQualifier> componentsPerQualifiers = searchInIndex(componentIndexQuery);
+    SuggestionsWsResponse searchWsResponse = toResponse(componentsPerQualifiers, warning);
+    writeProtobuf(searchWsResponse, wsRequest, wsResponse);
+  }
 
-    if (componentsPerQualifiers.isEmpty()) {
-      return Collections.emptyList();
+  private static String getWarning(String query) {
+    List<String> tokens = ComponentTextSearchFeature.split(query).collect(Collectors.toList());
+    if (tokens.stream().anyMatch(token -> token.length() < MINIMUM_NGRAM_LENGTH)) {
+      return SHORT_INPUT_WARNING;
     }
+    return null;
+  }
 
-    try (DbSession dbSession = dbClient.openSession(false)) {
-      return componentsPerQualifiers.stream().map(qualifier -> {
-
-        List<String> uuids = qualifier.getComponentUuids();
-        List<ComponentDto> componentDtos = dbClient.componentDao().selectByUuids(dbSession, uuids);
-        List<ComponentDto> sortedComponentDtos = Ordering.explicit(uuids)
-          .onResultOf(ComponentDto::uuid)
-          .immutableSortedCopy(componentDtos);
-
-        Map<String, String> organizationKeyByUuids = getOrganizationKeys(dbSession, componentDtos);
-
-        List<Component> results = sortedComponentDtos
-          .stream()
-          .map(dto -> dtoToComponent(dto, organizationKeyByUuids))
-          .collect(MoreCollectors.toList());
-
-        return Qualifier.newBuilder()
-          .setQ(qualifier.getQualifier())
-          .addAllItems(results)
-          .build();
-      }).collect(MoreCollectors.toList());
+  private List<ComponentsPerQualifier> getComponentsPerQualifiers(String more, ComponentIndexQuery.Builder queryBuilder) {
+    List<ComponentsPerQualifier> componentsPerQualifiers;
+    if (more == null) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Change this condition so that it does not always evaluate to "false" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2583)